### ISSUE TITLE
Line types customization

### ISSUE THREAD

**Issue by ghost** [LibreCAD/LibreCAD#1738]
Good morning, I read somewhere that in AutoCAD you can customize the line type simply through a text file that can be reached with the file explorer. Instead in LibreCAD, from what I understand, the line types are predefined, and to modify them you would have to recompile the source code ... but I'm not very familiar with these things. Could a change be made to the code so that LibreCAD works like AutoCAD? Thanks in advance for any help, have a nice day, Simone Aiello

**Comment by app4soft** [comment]
To not duplicate this issue, will add details from other feature request

# Support linetype Au

### Expected behavior

LibreCAD should support custom linetype definitions using setting path to AutoCAD linetype file (`.lin`).

Topic on Forum:

- https://forum.librecad.org/Creating-Custom-Linetypes-td4974845.html
- https://forum.librecad.org/Adding-custom-Line-Types-td5723518.html
- https://forum.librecad.org/Linetypes-aren-t-working-when-I-open-LibreCAD-dxf-files-with-AutoCAD-or-eDrawings-td5706867.html

### Observed behavior

LibreCAD has linetype definitions hardcoded in `rs_linetypepattern.cpp`:

- https://github.com/LibreCAD/LibreCAD/blob/master/librecad/src/lib/gui/rs_linetypepattern.cpp

### Steps to reproduce or sample file

AutoCAD linetype file (`.lin`) samples:

- https://github.com/tuffs/acad-lin
- https://github.com/sidhrika/MyLineTypes

LIN-file syntax highlighting for Notepad++:

- https://github.com/rddim/notepad-plus-plus-autocad

### Operating System and LibreCAD version info

All versions and all OS.

### Related issues

- Support hatch definition with AutoCAD hatch pattern files (`.pat`): https://github.com/LibreCAD/LibreCAD/issues/1651

**Comment by u2fly** [comment]
@dxli @lordofbikes

LibreCAD not fully supports standard ISO line types from DXF (produced by QCAD or other program) due to some unsupported line types, difference in metrics and line type substitution tables.

- ISO-128-2:2022 - https://standards.iteh.ai/catalog/standards/iso/ed38e2d0-844f-4eb5-97a9-8770b7a9acd6/iso-128-2-2022
  - Preview - https://cdn.standards.iteh.ai/samples/83355/10bb39d36fc34caeb80ecd25347ddb0c/ISO-128-2-2022.pdf
- QCAD's LIN file (ISO's lines metrics same to AutoCAD's LIN file) - https://github.com/qcad/qcad/blob/master/linetypes/metric/qcadiso.lin#L60-L93

Need to review:
- https://github.com/LibreCAD/LibreCAD/blob/master/librecad/src/lib/gui/rs_linetypepattern.cpp
- https://github.com/LibreCAD/LibreCAD/blob/master/librecad/src/lib/filters/rs_filterdxf.cpp
- https://github.com/LibreCAD/LibreCAD/blob/master/librecad/src/lib/filters/rs_filterdxfrw.cpp

> LIN-file support for LibreCAD would fix this and other possible issues.

(In table bellow, LibreCAD's line types in __Bold__ shows substitutes of ISO line types, prefaced with "__—/__" means there is no analogue in LibreCAD, suffixed with "__???__" means there is analogue in LibreCAD but not used for substitution.)

| ISO Type (* - not suppoted in LC) | DXF/LIN type code | LibreCAD substitution |
|--|:--:|--|
| Continious | `CONTINIOUS`<p>(`ACAD_ISO01W100`) | __CONTINIOUS__ |
| Dashed | `ACAD_ISO02W100` | __DASHED__ |
| * Dashed space | `ACAD_ISO03W100` | —/ __DASHED__ |
| Long-dashed dotted | `ACAD_ISO04W100` | __DASHDOTX2__ |
| Long-dashed double-dotted | `ACAD_ISO05W100` | __DIVIDEX2__ |
| * Long-dashed triplicate-dotted | `ACAD_ISO06W100` | —/ __CONTINIOUS__ |
| Dotted | `ACAD_ISO07W100` | __Dot__ |
| Long-dashed short-dashed | `ACAD_ISO08W100` | CENTER2??? / __CONTINIOUS__ |
| * Long-dashed double-short-dashed | `ACAD_ISO09W100` | —/ __CONTINIOUS__ |
| Dashed dotted | `ACAD_ISO10W100` | __DASHDOT__ 
| Double-dashed dotted | `ACAD_ISO11W100` | BORDER??? / __CONTINIOUS__ |
| Dashed double-dotted | `ACAD_ISO12W100` | __DIVIDE__ |
| * Double-dashed double-dotted | `ACAD_ISO13W100` | —/ __CONTINIOUS__ |
| * Dashed triplicate-dotted | `ACAD_ISO14W100` | —/ __BORDER__ |
| * Double-dashed triplicate-dotted | `ACAD_ISO15W100` | —/ __BORDER__ |

**Comment by ROMPEFUEGOS** [comment]
We would like to implement this, in small tested PRs like #2815 and #2823, and we would rather agree the design with you before writing product code.

**Why it matters beyond convenience.** CAM front-ends for cutting machines read the *linetype name* of each entity (group code 6) to choose the machining mode: stone and glass bridge saws, water-jet, plasma, press brakes (`BEND_UP`/`BEND_DOWN` with DASHED/CENTER), SigmaNEST-style nesting which maps by Layer/Colour/Line Type. Today `RS_FilterDXFRW::nameToLineType` returns `RS2::SolidLine` for any name outside the eight built-in families and `lineTypeToName` falls back to `CONTINUOUS`, so opening and saving such a drawing silently rewrites the information the machine needs. Since #2790 the LTYPE *records* survive in the saved file, but nothing references them any more.

**The plan** (draft, every anchor verified against master `f3aa61ac9`): https://github.com/ROMPEFUEGOS/LibreCAD/blob/feature/named-linetypes-engine/docs/plan_named_linetypes.md — same mould as `docs/plan_shp_native_filter.md`, and it opens with a "What we need from you" block, so the questions are the first thing you read.

In short: the document gets its own linetype table (`LC_LineTypeList`, shaped like `RS_LayerList`, seeded with the built-ins); `RS_Pen` carries the linetype **name** as identity with `RS2::LineType` kept as a derived "nearest built-in" cache (never renumbered) for settings, `.lcp` palettes, `$DIMLTYPE` and the plugin API; the DXF/DWG filters keep every name and every record. Later phases draw custom patterns from the table, list them in the pen selectors, and add a Linetypes dialog with `.lin` import/export. Only Phase 1 touches `rs_pen.h`.

**The one decision that unblocks Phase 1** is the pair Q1 + Q2:

1. **Q1** — document-owned table, with `.lin` files only *imported into* it? Or global `.lin` catalogues, the way QCAD seeds every new drawing, and the way `RS_PATTERNLIST`/`RS_FONTLIST` already resolve hatch patterns and fonts by name from files on disk? We propose the document table, because a linetype is not quite like a hatch pattern: DXF wants the LTYPE definition to travel inside the file, so the drawing (or a template) carries the contract with the machine and nothing has to be installed per workstation.
2. **Q2** — pen identity as name + cached enum, or a sentinel `RS2::CustomLine`? We propose the former; the sentinel still needs every consumer patched (`getPattern()` → nullptr, `findData()` → −1, `.lcp` validation, the `DPI::LineType` cast) and loses the name anyway.

Two more we would rather not guess on:

3. **Q3** — units for custom patterns when drawing: device millimetres exactly like the built-ins today (our proposal, no behaviour change for existing drawings), with `$LTSCALE`/drawing-unit scaling (#1476) as an optional later phase behind a document option?
4. **Q5** — a name referenced by an entity or layer with no LTYPE record in the file: synthesise a marker record and write it back? Without one the DWG writer resolves the name to handle 0 and the entity silently comes out ByLayer, and AutoCAD rejects a file whose entities reference an unregistered name.

Q4 and Q6–Q9 in the plan have defaults we will follow unless you say otherwise (keep alias names literal; complex text/shape linetypes kept opaque and rendered as their dash list, per @sand1024's point in #1917; the editor as a Drawing Preferences tab; no 2.2.1 backport; and we will not touch `dwg-dxf-verify.yml`'s path filter, which today skips the DXF/DWG test lanes for a PR that touches only `librecad/src/ui/**` or `librecad/src/lib/gui/render/**` — the Pixi builds and CodeQL still run, so such a PR is compiled but its behaviour is never exercised in CI).

Happy to split off the built-in metrics table move (out of `rs_filterdxfrw.cpp` into the engine) as its own preparatory PR if you prefer to review that first. We will not start Phase 1 until we hear from you.

**Comment by sand1024** [comment]
@ROMPEFUEGOS 

Well, support of linetypes depends on the level of linetype support. 

For straight lines - where individual drawing segment is just a combination of dots and lines - it's quite straightforward. 

Yet normal support of line types (where single line pattern step is not straight, like ziz-zag, or include some shape or text) - is actually quite a huge task. 

There are several area related to custom line types support: 

1) **Rendering**: Implement rendering of entities with custom linetypes (think about insulation line type, like https://forum.librecad.org/Batting-line-style-or-Insulation-drawing-td5724389.html, zig-zag or line with text). 
It is necessary to create fast oscillator for each line type segment step and rewrite painting logic for this (think about ellipse or splice drawn with custom linetype). 
Potentially this could be simplified if all painting will be line-approximation based, yet still - it's heavyweight.  
Moreover, as this is related to rendering - it's critical to ensure fast rendering, so things like caching and optimizations are needed there. 
Thus it might be that adding support of custom linetypes will require very deep rework of the entire rendering pipeline. 
Also, while I'm not fully sure in that - potentially it might be that custom line type could also affect snapping and hatching. 

2) **Persistence**: There are standard definitions for line types as well as how they are stored in the document - they could be found in DXF format specification and they are supported by AutoCAD or so. It's necessary to pay close attention to existing standards and interoperability.  Here support of custom line types parsing will require support of line type patterns, SHP format and so on. And, corresponding serialization, for sure. As for external .lin files - I don't remember exactly, yet it seems that they are common patterns for lines that are still later embedded into drawing. 

3) **Data** - of course, line types should be supported in all places where RS2::LineType is used. 

4) **UI** rework controls for line type selection and displaying, plus custom line types editor. 

As for me, complete implementation of data, persistence and UI components is rather useless without proper support of reliable and fast rendering  (item 1).

All preparation phases are just require efforts, yet, in general, they are trivial (except careful support of DXF format for reading/writing and maintaining interoperability).

**Comment by ROMPEFUEGOS** [comment]
Thank you for the taxonomy. We took the rendering half seriously rather than arguing, and it changed what we are proposing.

**We profiled your pipeline and spiked the change.** A repeatable headless benchmark; four full patched builds to price the identity change itself; then, for the fixes below, pen and QPen call counts under LD_PRELOAD and gdb breakpoint counts across six separately built arms. Two things came out of it that matter more than our feature.

**1. Three pre-existing defects in the dashed path, which we are sending as three separate PRs, in this order.**

- `rsToQDashPattern()` dereferences `RS_LineTypePattern::getPattern()` without a null check. `RS2::LineTypeUnchanged` has no table entry, and neither does an out-of-range value, so a stale settings integer reaching a grid, snap or overlay pen is a crash. That one is a one-line guard plus a regression case: #2852.
- `RS_Painter::setPen()`'s `Qt::CustomDashLine` branch returns before the `bool changed` block the solid branch uses, so a dashed pen is rebuilt and reinstalled for every entity. Counted on a 200 000-entity drawing: all solid costs 1 `QPainter::setPen()` call, 0 `QPen` constructions and 0 dash-pattern allocations; all dashed costs 200 000 of each. The fix keys on what actually determines the pattern — line type, screen width, resolution — instead of expanding and comparing the `QList`, so it also helps the case we were most worried about: on a **mixed** drawing whose consecutive entities rotate through four line types, `QPen` constructions go **150 000 → 0** while `setDashPattern` stays at 150 000, i.e. it discriminates rather than skipping blindly. Counted at six arms — master, the three commits proposed here, the intermediate commit of the third, and a superseded attempt kept as a control — each relinked against its own library, with the library's checksum recorded, so no arm can be a stale binary. #2853.
- `RS_Pen::isSameAs()` compares the running dash offset, which `updateDashOffset()` moves for every entity drawn, against a cached pen whose own offset is always 0. Counted with a gdb breakpoint over one pass of 2000 identical-pen lines: all solid gives 1 `RS_Painter::setPen()` call, one dashed line **last** gives 2, one dashed line **first** gives **2000**. A single dashed entity at the front costs the cache for the rest of the frame. #2854.

**2. What we will not claim.** We cannot demonstrate a wall-clock win for any of it on our harness, and we would rather say so than quote a number you would be right to distrust. The timing attempt was a separate, smaller run than the counts above: three arms rebuilt from the real commits, stripped, interleaved, with an untouched all-solid control in every round. The control itself moves between +0.8% and −1.5% depending on the comparison, which is the same order as the effects, and runs at two drawing sizes disagree in sign. So the case for these three is a crash, and allocations and calls removed, counted rather than timed. If you have a harness you trust for this, we will run whatever you ask on it.

**On the feature itself, our scope is only the half you called straightforward.** Dash lists, the acad.lin shapes. Patterns embedding a shape or text stay out: their `DRW_LType` records are preserved opaque so a round-trip does not destroy them, and they draw as the dash part only, which is the line you took in #1917.

Two smaller findings while we were in there, offered rather than pressed. On the snapping and hatching worry: nothing in either reads `RS_LineTypePattern`. The pattern table has exactly two consumers, `rsToQDashPattern` and the pass-through in `RS_GraphicView`, and the dash phase lives entirely inside `RS_Painter`. `rs_snapper.cpp` builds its indicator pen from a settings integer and never reads an entity pen, and `rs_hatch.cpp` resolves its pattern through `RS_PATTERNLIST` and names no line type anywhere in its 580 lines, so a dash-only feature cannot reach them. One qualification, in the direction that costs us rather than flatters us: a hatch propagates its own pen to each child and draws them one at a time, so a **dashed** hatch does re-enter the painter per child. That makes it a beneficiary of the second fix above, not a case the fix leaves alone. And Qt bounds any custom-pattern design in two places worth knowing: `QDashStroker::repetitionLimit()` is a hard 10000 periods, past which the line is drawn **solid** (bisected at exactly 10000.0 at three pen widths), and a pen wider than 1 leaves the raster fast path, where the same 50 000 dashed lines went from 135 ms to 817 ms.

The plan, with every anchor verified and the measurements attached: https://github.com/ROMPEFUEGOS/LibreCAD/blob/77e3a75b729d2d8c923659accc5e675e61932e98/docs/plan_named_linetypes.md

We are not asking you to accept the document-owned table yet. We would like to know whether the three fixes above are welcome, and in what order you would want them.

**Comment by ROMPEFUEGOS** [comment]
@sand1024 @dxli — an update, and one question that needs an answer before we write engine code.

**The rendering half came first, as you asked.** Profiling your pipeline turned up four pre-existing defects in the dashed path, and all four are now in master: #2852 (a null dereference), #2853 (the dashed `QPen` rebuilt for every entity), #2854 (the dash offset compared against a cached pen whose own offset is always zero) and #2879 (draft and scaled pens answering for each other in the same cache). Two of them @dxli reshaped on the way in, which improved them. We are not asking for anything back for that work; it stands on its own.

**Our scope is still only the half you called straightforward**: dash lists. Patterns carrying a shape or text stay opaque — preserved through a round trip, drawn as their dash part — which is the line you took in #1917, and the staging you proposed there yourself: the simple form first.

**The first step is a refactor we offered on 12 September**, and it is open now as #2880: `nameToLineType()` and `lineTypeToName()` move from `RS_FilterDXFRW` into the engine, so `LC_DimStyle` stops including the DXF filter header to translate three header variables. No behaviour change, no new state.

**The question.** Phase 1 gives the document its own linetype table and keeps every name through DXF and DWG. Two decisions shape it:

- **Q1** — the table belongs to the document, with `.lin` files only imported into it? Or global `.lin` catalogues, the way hatch patterns and fonts already resolve by name from disk?
- **Q2** — the pen carries the name (with the enum kept as a derived cache), or a sentinel `RS2::CustomLine`?

We propose the document-owned table and the name as identity, with a 16-bit interned id in the pen so the pen does not grow and comparisons stay integer comparisons. Both are reversible while nothing in the UI depends on them.

Unless one of you would rather we took another route, we will start on that basis after **1 October**. If the answer is "not yet", say so and we will keep the work on the fork.
